core/src/state/world_state.rs

```diff
@@ -65,44 +66,134 @@ impl StateHandle {
             .write()
             .expect("Could not acquire world_state lock.")
     }
+}
+
+/*
+We want a thing that waits until notified, and importantly waits
+for that notify on all of its replicas.
+The default notifier from tokio does not behave the way we need
+since, if multiple clones are made, notify_one only causes the first
+to return immediately when it invokes the notification and awaits.
+notify_waiters does not cause even the first one to do that.
+Therefore, we implement our own notifier that can be closed and
+will return immediately on all its dependent notifies.
+*/
+#[derive(Debug)]
+pub struct ClosableNotify {
+    notifier: Arc<Sender<()>>,
+    rx: Receiver<()>,
+}
+
+impl ClosableNotify {
+    fn new() -> Self {
+        let (tx, rx) = tokio::sync::broadcast::channel(128);
+        Self {
+            notifier: Arc::new(tx),
+            rx,
+        }
+    }
 
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
+    pub async fn notified(&mut self) {
+        let _ = self.rx.recv().await;
+    }
+
+    fn notify_waiters(&self) {
+        let _ = self.notifier.send(());
+    }
+}
+
+impl Clone for ClosableNotify {
+    fn clone(&self) -> Self {
+        Self {
+            notifier: self.notifier.clone(),
+            rx: self.notifier.subscribe(),
         }
-        let receiver = { self.write_state().get_listener(sequence) };
-        receiver.notified().await;
     }
 }
 
+pub trait DebugConditionT: Sync + Send + Fn(&WorldState) -> bool {}
```

Comment:
please let your eyes glaze over this whole song and dance. it's just a workaround for the fact that Fn doesn't implement debug.